hw/misc/pcie-testdev.c

```diff
@@ -0,0 +1,202 @@
+
+typedef struct DMARegs {
+    dma_addr_t srcaddr;
+    dma_addr_t dstaddr;
+    dma_addr_t size;
+    dma_addr_t doorbell;
+    dma_addr_t status;
+    dma_addr_t busy_delay;
+} DMARegs;
+
+typedef struct PCIeTestDevState {
+    PCIDevice parent_obj;
+    MemoryRegion membar;
+    DMARegs regs;
+    bool aggressive;
+    bool aggression_delay; /* delay between aggressive DMA transfers */
+} PCIeTestDevState;
+
+#define PCI_EXPRESS_TEST_DEV(obj) \
+    OBJECT_CHECK(PCIeTestDevState, (obj), TYPE_PCIE_TEST_DEV)
+
+static Property props[] = {
+    DEFINE_PROP_BOOL("aggressive", PCIeTestDevState, aggressive, false),
+    DEFINE_PROP_UINT64("aggression_delay_ms", PCIeTestDevState, aggression_delay, false),
+    DEFINE_PROP_UINT64("srcaddr", PCIeTestDevState, regs.srcaddr, false),
+    DEFINE_PROP_UINT64("dstaddr", PCIeTestDevState, regs.dstaddr, false),
+    DEFINE_PROP_UINT64("size", PCIeTestDevState, regs.size, false),
+    DEFINE_PROP_UINT64("busy_delay_ms", PCIeTestDevState, regs.busy_delay, false),
+    DEFINE_PROP_END_OF_LIST()
+};
+
+static uint64_t
+pcie_testdev_read(void *opaque, hwaddr addr, unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        return s->regs.srcaddr;
+        break;
+    case DSTADDR:
+        return s->regs.dstaddr;
+        break;
+    case SIZE:
+        return s->regs.size;
+        break;
+    case STATUS:
+        return s->regs.status;
+        break;
+    case BUSY_DELAY:
+        return s->regs.busy_delay;
+    default:
+        return 0x0;
+    }
+}
+
+static void
+enable_bus_master(PCIeTestDevState *s){
+    pci_set_word(s->parent_obj.config + PCI_COMMAND,
+                 s->parent_obj.config[PCI_COMMAND] | PCI_COMMAND_MASTER);
+}
+
+static void
+dma_transfer(PCIeTestDevState *s){
+    s->regs.status = 0x01;
+    printf("DMA Transfer from source addr %p to dest addr %p, length %lu\n", (void *)s->regs.srcaddr, (void *)s->regs.dstaddr, s->regs.size);
+    uint16_t *buf = g_malloc(s->regs.size);
+    printf("Read from source, STATUS: %d\n", pci_dma_read(&s->parent_obj, s->regs.srcaddr, buf, s->regs.size));
+    printf("Write to dest, STATUS: %d\n", pci_dma_write(&s->parent_obj, s->regs.dstaddr, buf, s->regs.size));
+    g_free(buf);
+    s->regs.status = 0x00;
+
+    if (s->aggressive) {
+        QEMUTimer *dma_timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+        timer_mod(dma_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + s->aggression_delay);
+    }
+}
+
+static void
+pcie_testdev_write(void *opaque, hwaddr addr, uint64_t val,
+                   unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        s->regs.srcaddr = val;
+        break;
+    case DSTADDR:
+        s->regs.dstaddr = val;
+        break;
+    case SIZE:
+        s->regs.size = val;
+        break;
+    case DOORBELL:
+        s->regs.doorbell = val;
+        if (val & 0x01 && s->regs.size) {
```

Comment:
I'd recommend parentheses when you have both & and && together. The precedence rules sometimes surprise me, and I keep forgetting which ones are surprising.

Reply:
Perhaps also verify that the status register does not indicate busy -- you shouldn't re-arm the timer if it has already been set.

Reply:
Added a check for the status, and LOG_GUEST_ERROR if an attempt is made while device is busy